### ISSUE TITLE
[py2py3] Check unittests from WMCore/Credential in py3

### ISSUE THREAD

**Issue by mapellidario** [dmwm/WMCore#10529]
#### Intro

We need to make sure that the work done in the migration slices
is py3 compatible. In particular, we should make sure that the
unittests that are involved in each slice do not fail.

The schedule for this validation / bug-fixing campaign is kept at #10422

#### Details

This issue is about checking that the unittests from WMCore/Credential,
issue #10007 and PR #10008, do not fail.
A non-exhaustive list of the ut that need to pass is

```
test/python/WMCore_t/Credential_t/Proxy_t.py
test/python/WMCore_t/Credential_t/SimpleMyProxy_t.py

```

We should make sure that the PR that will close this issue, also takes care of 
- [ ] #10483

**Comment by mapellidario** [comment]
Both in `gitlab-registry.cern.ch/cmsdocks/dmwm:wmcorepy3_tests`, gitlab-registry.cern.ch/cmsdocks/dmwm:wmcore_tests` and my VM with py2 and oracle many unittests in `test/python/WMCore_t/Credential_t/*` are failing with something that feels related to

```plaintext
[dmwm@6815d86daa5c WMCore]$ voms-proxy-info
voms-proxy-info: symbol lookup error: voms-proxy-info: undefined symbol: _ZN8vomsdataC1ESsSs
```

@amaltaro am i the only one to have these problems? Should we change our environment?

**Comment by amaltaro** [comment]
I've seen such errors in the past, it comes from a mixture of libraries where one of them breaks our VOMS utils.
Perhaps we could try assigning it to either Todor or Erik? Or I can try looking into it tomorrow.

**Comment by mapellidario** [comment]
Sounds good, i will remove myself from the assignees! Thanks Alan!

**Comment by amaltaro** [comment]
@todor-ivanov or @goughes would you be able to investigate this issue today/tomorrow? If needed, compare it against the actual WMAgent nodes. 

If you can spare some time on it, please assign it to yourself, otherwise I will work on it on Thursday.

**Comment by goughes** [comment]
@amaltaro I could put some time into this tomorrow if @todor-ivanov hasn't already starting working on it.

**Comment by goughes** [comment]
From my testing, `voms-proxy-*` commands fail once you source the wmagent environment in either py2 `env_unittest.sh` or the py3 `env_unittest_py3.sh`. 

```
source $INSTALL_DIR/current/apps/wmagentpy3/etc/profile.d/init.sh
```

@amaltaro One thing that is peculiar to me... I do not see these `Credential_t` tests here: https://cmssdt.cern.ch/dmwm-jenkins/job/DMWM-WMCore-UnitTests/lastCompletedBuild/testReport/

Does Jenkins even run these?

**Comment by goughes** [comment]
Some additional info: 

I took a look on my test agent and found that there is a difference in the `voms-clients` package between what's in the Docker image and what is on the agent.

The Docker image uses `voms-clients-cpp` which gets messed up once LD_LIBRARY_PATH changes when init'ing the agent env.
```
[dmwm@b818781dc4b7 WMCore]$ rpm -q voms-clients-cpp
voms-clients-cpp-2.0.16-1.el7.x86_64
```

The agents have both the cpp and java versions:
```
[egough@vocms0264 ~]$ rpm -q voms-clients-cpp
voms-clients-cpp-2.0.16-1.el7.x86_64
[egough@vocms0264 ~]$ rpm -q voms-clients-java
voms-clients-java-3.3.2-1.el7.noarch
```

The Java client is used by default on the agents:
```
[egough@vocms0264 ~]$ which voms-proxy-init
/usr/bin/voms-proxy-init
[egough@vocms0264 ~]$ ls -l /usr/bin/voms-proxy-init
lrwxrwxrwx. 1 root root 33 Apr 12 14:11 /usr/bin/voms-proxy-init -> /etc/alternatives/voms-proxy-init
[egough@vocms0264 ~]$ ls -l /etc/alternatives/voms-proxy-init
lrwxrwxrwx. 1 root root 25 Apr 12 14:11 /etc/alternatives/voms-proxy-init -> /usr/bin/voms-proxy-init3
[egough@vocms0264 ~]$ cat /usr/bin/voms-proxy-init3
#!/bin/sh
VOMS_CLIENTS_JAVA_OPTIONS=${VOMS_CLIENTS_JAVA_OPTIONS:-"-XX:+UseSerialGC -Xmx16m"}
java $VOMS_CLIENTS_JAVA_OPTIONS -cp $(build-classpath voms-clients-java voms-api-java canl-java bcpkix bcprov commons-cli commons-io) org.italiangrid.voms.clients.VomsProxyInit "$@"
```

I think using the Java client in the unittest Docker image would solve at least this initial symbol lookup problem.

**Comment by amaltaro** [comment]
If I look at this unit test module:
https://github.com/dmwm/WMCore/blob/master/test/python/WMCore_t/Credential_t/Proxy_t.py

it looks like all the unit tests are marked as `integration`, which means, they are not executed by Jenkins (usually because we have problems with those tests in Jenkins, or because they are simply broken...). I did not check the other two modules though.

@goughes this PR matches the description you have provided:
https://github.com/dmwm/WMCore/pull/9649
and as you can see, it's not the first time this incompatibility turns to be a problem for us.

Can you please check the following:
* whether all those unit tests are marked as `integration` under the `Credential_t` package
* how those run from your Py2 docker container (in terms of # of failures/success)
* how those run from your Py3 docker container (in terms of # of failures/success)

depending on the outcome, I would suggest to either a) close this one and make a separate issue to sort those unit tests; or b) repurpose this GH issue to fix those unit tests (outside the Py3 migration scope).

**Comment by goughes** [comment]
All the tests are marked as `integration`.

### test/python/WMCore_t/Credential_t/Proxy_t.py

I ran these with AND without sourcing the wmagent environment. Errors/failures were the same for each run.

py2: `FAILED (errors=2, failures=10)`
py3: `FAILED (errors=4, failures=8)`

Some py3 errors that may be of concern:

```
======================================================================           
ERROR: Test if the checkAttribute  method checks correctly the attributes validity.
----------------------------------------------------------------------           
Traceback (most recent call last):                                               
  File "/home/dmwm/wmcore_unittest/WMCore/test/python/WMCore_t/Credential_t/Proxy_t.py", line 167, in testCheckAttribute
    valid = self.proxy.checkAttribute( )                                         
  File "/home/dmwm/wmcore_unittest/WMCore/src/python/WMCore/Credential/Proxy.py", line 374, in checkAttribute
    if not re.compile(r"^" + proxyDetails).search(attribute[0]):                 
TypeError: cannot use a string pattern on a bytes-like object                    
-------------------- >> begin captured logging << --------------------           
ProxyTest: DEBUG: Executing :                                                    
 command : voms-proxy-info -fqan -file /tmp/x509up_u1000                         
 output : b''                                                                    
 error: b"\nCouldn't find a valid proxy.\n\n"                                    
 retcode : 1                                                                     
--------------------- >> end captured logging << ---------------------         
```

```
======================================================================           
ERROR: Test if getTimeLeft method returns correctly the proxy time left.         
----------------------------------------------------------------------           
Traceback (most recent call last):                                               
  File "/home/dmwm/wmcore_unittest/WMCore/test/python/WMCore_t/Credential_t/Proxy_t.py", line 91, in testGetUserCertEnddate
    daysleft = self.proxy.getUserCertEnddate()                                   
  File "/home/dmwm/wmcore_unittest/WMCore/src/python/WMCore/Credential/Proxy.py", line 234, in getUserCertEnddate
    timeleft = self.getUserCertTimeLeft(openSSL)                                 
  File "/home/dmwm/wmcore_unittest/WMCore/src/python/WMCore/Credential/Proxy.py", line 199, in getUserCertTimeLeft
    out = out.split('notAfter=')[1]                                              
TypeError: a bytes-like object is required, not 'str'                            
-------------------- >> begin captured logging << --------------------           
ProxyTest: DEBUG: Executing :                                                    
 command : openssl x509 -noout -in /home/dmwm/certs/servicecert.pem -dates          
 output : b'notBefore=Jun  2 14:38:39 2021 GMT\nnotAfter=Jun  3 02:43:39 2021 GMT\n'
 error: b''                                                                      
 retcode : 0                                                                     
--------------------- >> end captured logging << ---------------------           
```

```
======================================================================           
ERROR: _testGetUserName_                                                         
----------------------------------------------------------------------           
Traceback (most recent call last):                                               
  File "/home/dmwm/wmcore_unittest/WMCore/test/python/WMCore_t/Credential_t/Proxy_t.py", line 156, in testGetUserName
    user = self.proxy.getUserName( )                                             
  File "/home/dmwm/wmcore_unittest/WMCore/src/python/WMCore/Credential/Proxy.py", line 309, in getUserName
    for cname in subject.split('/'):                                             
TypeError: a bytes-like object is required, not 'str'                            
-------------------- >> begin captured logging << --------------------           
ProxyTest: DEBUG: Executing :                                                    
 command : voms-proxy-info -file /tmp/x509up_u1000 -identity                     
 output : b''                                                                    
 error: b"\nCouldn't find a valid proxy.\n\n"                                    
 retcode : 1                                                                     
--------------------- >> end captured logging << ---------------------           
```
### test/python/WMCore_t/Credential_t/MyProxy_t.py

I ran these when sourcing the wmagent environment.

All 6 tests fail in the `setUp` method on both py2 and py3 with:
```
Traceback (most recent call last):
  File "/home/dmwm/wmcore_unittest/WMCore/test/python/WMCore_t/Credential_t/MyProxy_t.py", line 47, in setUp
    self.proxy = Proxy( self.dict )
  File "/home/dmwm/wmcore_unittest/WMCore/src/python/WMCore/Credential/Proxy.py", line 161, in __init__
    raise CredentialException('voms-proxy-info command not found')
```
### test/python/WMCore_t/Credential_t/SimpleMyProxy_t.py

I ran these when sourcing the wmagent environment.

Fails in py2 with:
```
======================================================================
ERROR: Test if it is possible to retrieve a proxy from an existing myproxy
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dmwm/wmcore_unittest/WMCore/test/python/WMCore_t/Credential_t/SimpleMyProxy_t.py", line 68, in testgetMyProxy
    keyfile=SERVERKEY, certfile=SERVERCERT, lifetime=43200)
  File "/home/dmwm/wmcore_unittest/WMCore/src/python/WMCore/Credential/SimpleMyProxy.py", line 214, in logonRenewMyProxy
    sslctx = myproxy_ctx(certfile if certfile else self.serverCert, keyfile if keyfile else self.serverKey)
  File "/home/dmwm/wmcore_unittest/WMCore/src/python/WMCore/Credential/SimpleMyProxy.py", line 44, in myproxy_ctx
    ctx.use_certificate_chain_file(certfile)
  File "/home/dmwm/unittestdeploy/wmagent/1.4.9.pre3/sw/slc7_amd64_gcc630/external/py2-pyOpenSSL/18.0.0-comp3/lib/python2.7/site-packages/OpenSSL/SSL.py", line 913, in use_certificate_chain_file
    _raise_current_error()
  File "/home/dmwm/unittestdeploy/wmagent/1.4.9.pre3/sw/slc7_amd64_gcc630/external/py2-pyOpenSSL/18.0.0-comp3/lib/python2.7/site-packages/OpenSSL/_util.py", line 54, in exception_from_error_queue
    raise exception_type(errors)
Error: [('system library', 'fopen', 'No such file or directory'), ('BIO routines', 'FILE_CTRL', 'system lib'), ('SSL routines', 'SSL_CTX_use_certificate_chain_file', 'system lib')]

```

Fails in py3 with:
```
======================================================================
ERROR: Failure: RuntimeError (You are linking against OpenSSL 1.0.2, which is no longer supported by the OpenSSL project. To use this version of cryptography you need to upgrade to a newer version of OpenSSL. For this version only you can also set the environment variable CRYPTOGRAPHY_ALLOW_OPENSSL_102 to allow OpenSSL 1.0.2.)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dmwm/unittestdeploy/wmagent/1.4.9.pre3-comp/sw/slc7_amd64_gcc630/external/py3-nose/1.3.7/lib/python3.8/site-packages/nose/failure.py", line 39, in runTest
    raise self.exc_val.with_traceback(self.tb)
  File "/home/dmwm/unittestdeploy/wmagent/1.4.9.pre3-comp/sw/slc7_amd64_gcc630/external/py3-nose/1.3.7/lib/python3.8/site-packages/nose/loader.py", line 416, in loadTestsFromName
    module = self.importer.importFromPath(
  File "/home/dmwm/unittestdeploy/wmagent/1.4.9.pre3-comp/sw/slc7_amd64_gcc630/external/py3-nose/1.3.7/lib/python3.8/site-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/home/dmwm/unittestdeploy/wmagent/1.4.9.pre3-comp/sw/slc7_amd64_gcc630/external/py3-nose/1.3.7/lib/python3.8/site-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/home/dmwm/unittestdeploy/wmagent/1.4.9.pre3-comp/sw/slc7_amd64_gcc630/external/python3/3.8.2-comp/lib/python3.8/imp.py", line 234, in load_module
    return load_source(name, filename, file)
  File "/home/dmwm/unittestdeploy/wmagent/1.4.9.pre3-comp/sw/slc7_amd64_gcc630/external/python3/3.8.2-comp/lib/python3.8/imp.py", line 171, in load_source
    module = _load(spec)
  File "<frozen importlib._bootstrap>", line 702, in _load
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 783, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/dmwm/wmcore_unittest/WMCore/test/python/WMCore_t/Credential_t/SimpleMyProxy_t.py", line 15, in <module>
    from WMCore.Credential.SimpleMyProxy import SimpleMyProxy
  File "/home/dmwm/wmcore_unittest/WMCore/src/python/WMCore/Credential/SimpleMyProxy.py", line 20, in <module>
    from OpenSSL import SSL, crypto
  File "/home/dmwm/unittestdeploy/wmagent/1.4.9.pre3-comp/sw/slc7_amd64_gcc630/external/py3-pyOpenSSL/18.0.0/lib/python3.8/site-packages/OpenSSL/__init__.py", line 8, in <module>
    from OpenSSL import crypto, SSL
  File "/home/dmwm/unittestdeploy/wmagent/1.4.9.pre3-comp/sw/slc7_amd64_gcc630/external/py3-pyOpenSSL/18.0.0/lib/python3.8/site-packages/OpenSSL/crypto.py", line 16, in <module>
    from OpenSSL._util import (
  File "/home/dmwm/unittestdeploy/wmagent/1.4.9.pre3-comp/sw/slc7_amd64_gcc630/external/py3-pyOpenSSL/18.0.0/lib/python3.8/site-packages/OpenSSL/_util.py", line 6, in <module>
    from cryptography.hazmat.bindings.openssl.binding import Binding
  File "/home/dmwm/unittestdeploy/wmagent/1.4.9.pre3-comp/sw/slc7_amd64_gcc630/external/py3-cryptography/3.2.1/lib/python3.8/site-packages/cryptography/hazmat/bindings/openssl/binding.py", line 222, in <module>
    _verify_openssl_version(Binding.lib)
  File "/home/dmwm/unittestdeploy/wmagent/1.4.9.pre3-comp/sw/slc7_amd64_gcc630/external/py3-cryptography/3.2.1/lib/python3.8/site-packages/cryptography/hazmat/bindings/openssl/binding.py", line 182, in _verify_openssl_version
    raise RuntimeError(
RuntimeError: You are linking against OpenSSL 1.0.2, which is no longer supported by the OpenSSL project. To use this version of cryptography you need to upgrade to a newer version of OpenSSL. For this version only you can also set the environment variable CRYPTOGRAPHY_ALLOW_OPENSSL_102 to allow OpenSSL 1.0.2.

----------------------------------------------------------------------
Ran 1 test in 0.000s

```

**Comment by amaltaro** [comment]
Thanks for this investigation, Erik! I would suggest to set this issue aside for now (move it back to the ToDo column) and work on other issues created from the python3 migration (this was anyways already broken). Some of these errors might be a real problem when we deploy things, but it should be quite isolated (AgentStatusWatcher).

If you agree, please go ahead. @goughes

**Comment by goughes** [comment]
Sounds good. Moving back to ToDo.

**Comment by goughes** [comment]
I re-tested this with Docker. Here are the steps I had to take to get past the voms-client issues:

* Start container as root: add `-u 0` to the docker run command
* Install voms-clients-java: `yum -y install voms-clients-java`
* Edit `/etc/yum.repos.d/wlcg-centos7.repo` and set `gpgcheck=0`
* Install VOMS info: `yum -y install wlcg-voms-cms`
* Switch to the dmwm user: `su - dmwm`
* Copy a valid proxy to `/tmp/x509up_u1000` 
* Source unittest env, run tests, etc.

I think errors from `test/python/WMCore_t/Credential_t/Proxy_t.py` will need to be fixed. I can do a PR for that.

@amaltaro @mapellidario

**Comment by goughes** [comment]
So I just noticed that `execute_command` was a WMCore function and I can simply do the bytes/string checking there, which will also fix many things related to the MyProxy functions. I will update my PR.

**Comment by goughes** [comment]
With the changes from #10577 I can run the whole set of `Proxy_t` unittests with 0 errors and 3 consistent failures. Two tests fail (I think?) because I cannot set a "group" attribute on my proxy: `testGetUserCertEnddate` and `testUserGroupInProxy`. `testGetUserCertEnddate` fails because my cert expiration date isn't 58 days from now. I do not understand that test.

```
======================================================================
FAIL: Test getAttributeFromProxy method.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dmwm/wmcore_unittest/WMCore/test/python/WMCore_t/Credential_t/Proxy_t.py", line 231, in attributes
    self.assertTrue(self.proxy.group, 'No group set. Testing incomplete.')
AssertionError: '' is not true : No group set. Testing incomplete.

======================================================================
FAIL: Test if getTimeLeft method returns correctly the proxy time left.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dmwm/wmcore_unittest/WMCore/test/python/WMCore_t/Credential_t/Proxy_t.py", line 96, in testGetUserCertEnddate
    self.assertEqual(daysleft, 58) #set this as the number of days left in .globus/usercert.pem
AssertionError: 0 != 58
-------------------- >> begin captured logging << --------------------
ProxyTest: DEBUG: Executing : 
 command : openssl x509 -noout -in /home/dmwm/certs/servicecert.pem -dates
 output : notBefore=Jun  7 15:46:02 2021 GMT
notAfter=Jun  8 03:51:02 2021 GMT

 error: 
 retcode : 0
--------------------- >> end captured logging << ---------------------

======================================================================
FAIL: Test if getUserAttributes method returns correctly the user group.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dmwm/wmcore_unittest/WMCore/test/python/WMCore_t/Credential_t/Proxy_t.py", line 211, in testUserGroupInProxy
    self.assertTrue(self.proxy.group, 'No group set. Testing incomplete.')
AssertionError: '' is not true : No group set. Testing incomplete.

----------------------------------------------------------------------
Ran 16 tests in 192.074s

FAILED (failures=3)
```

FYI @amaltaro @mapellidario

**Comment by amaltaro** [comment]
@goughes Erik, I'm not sure what exactly is the question here. But if you fixed 13 uni tests, and 3 keep failing, then I'd suggest to remove the `integration` flag on those 13 unit tests and see how those go in our jenkins setup. If it succeeds, then that's a reasonable improvement IMO.

**Comment by goughes** [comment]
@amaltaro No question really, just that I think #10577 should be merged. I think this will help Stefano's Proxy issues in https://github.com/dmwm/WMCore/issues/10483. 

If removing the integration tag will make the `Proxy_t` Jenkins tests run inside the container, then I don't think they will pass, as I had to do all the additional stuff from https://github.com/dmwm/WMCore/issues/10529#issuecomment-856175562 to get them to work locally.

**Comment by amaltaro** [comment]
Gotcha! Let's do it! Thanks, Erik.

**Comment by goughes** [comment]
I spent some more time with the `SimpleMyProxy_t` tests and I end up with the same failures on py2 and py3.

```
======================================================================
ERROR: Test if it's possible to get myproxy information
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dmwm/wmcore_unittest/WMCore/test/python/WMCore_t/Credential_t/SimpleMyProxy_t.py", line 56, in testgetMyProxyInfo
    myproxyinfo = self.simplemyproxy.checkMyProxy(username=USERNAME, myproxyserver=MYPROXYSERVER, myproxyport=MYPROXYPORT,
  File "/home/dmwm/wmcore_unittest/WMCore/src/python/WMCore/Credential/SimpleMyProxy.py", line 196, in checkMyProxy
    myproxy = myproxy_client(sslctx, 'info', username if username else self.userName, logger=self.logger,
  File "/home/dmwm/wmcore_unittest/WMCore/src/python/WMCore/Credential/SimpleMyProxy.py", line 132, in myproxy_client
    conn.write('0')
  File "/home/dmwm/unittestdeploy/wmagent/1.4.9.pre6-comp2/sw/slc7_amd64_gcc630/external/py3-pyOpenSSL/18.0.0/lib/python3.8/site-packages/OpenSSL/SSL.py", line 1729, in send
    self._raise_ssl_error(self._ssl, result)
  File "/home/dmwm/unittestdeploy/wmagent/1.4.9.pre6-comp2/sw/slc7_amd64_gcc630/external/py3-pyOpenSSL/18.0.0/lib/python3.8/site-packages/OpenSSL/SSL.py", line 1639, in _raise_ssl_error
    _raise_current_error()
  File "/home/dmwm/unittestdeploy/wmagent/1.4.9.pre6-comp2/sw/slc7_amd64_gcc630/external/py3-pyOpenSSL/18.0.0/lib/python3.8/site-packages/OpenSSL/_util.py", line 54, in exception_from_error_queue
    raise exception_type(errors)
OpenSSL.SSL.Error: [('SSL routines', 'SSL3_GET_RECORD', 'wrong version number')]
```
This error makes me think the code tries to connect to `myproxy.cern.ch` over SSLv3, but the server does not support it. 

https://github.com/dmwm/WMCore/blob/1bb206d606c4acc09c29a19c08a057a51de2c235/src/python/WMCore/Credential/SimpleMyProxy.py#L43

It seems hard coded to use this, so I don't know how or if it works when clients (like CRAB) try to use this module. :thinking:

**Comment by amaltaro** [comment]
@goughes Erik, thanks for looking into this issue once again.

Given that this fix has been made and merged:  #10577 I'm inclined to assume this work as done and the issue fixed. Once CRAB can actually start testing/using this, we will know whether there are other problems that were not completely spotted with our current setup. What do you think?

**Comment by goughes** [comment]
@amaltaro This sounds good to me. It's hard to know how long it has been since the `SimpleMyProxy_t` test was checked. Let's see if any issues pop up during CRAB testing. Closing the issue.
